internal/librarian/createreleasepr.go

```diff
@@ -308,24 +310,29 @@ func generateReleaseCommitForEachLibrary(state *commandState, inputDirectory str
 }
 
 func formatReleaseNotes(commitMessages []*CommitMessage) string {
-	features := []string{}
-	docs := []string{}
-	fixes := []string{}
+	features := make(map[string]bool)
+	docs := make(map[string]bool)
+	fixes := make(map[string]bool)
 
-	// TODO(https://github.com/googleapis/librarian/issues/549): deduping (same message across multiple commits)
 	// TODO(https://github.com/googleapis/librarian/issues/547): perhaps record breaking changes in a separate section
 	// TODO(https://github.com/googleapis/librarian/issues/550): include backlinks, googleapis commits etc
 	for _, commitMessage := range commitMessages {
-		features = append(features, commitMessage.Features...)
-		docs = append(docs, commitMessage.Docs...)
-		fixes = append(fixes, commitMessage.Fixes...)
+		for _, feature := range commitMessage.Features {
+			features[feature] = true
+		}
+		for _, doc := range commitMessage.Docs {
+			docs[doc] = true
+		}
+		for _, fix := range commitMessage.Fixes {
+			fixes[fix] = true
+		}
 	}
 
 	var builder strings.Builder
 
-	maybeAppendReleaseNotesSection(&builder, "New features", features)
-	maybeAppendReleaseNotesSection(&builder, "Bug fixes", fixes)
-	maybeAppendReleaseNotesSection(&builder, "Documentation improvements", docs)
+	maybeAppendReleaseNotesSection(&builder, "New features", slices.Collect(maps.Keys(features)))
```

Comment:
Sort for determinism.
Or if order matters, you need to use a slice, and use a map only to dedup.

Reply:
Rewrote to preserve the initial ordering (we use the map to only decide if we have seen the entry before)

Added a TODO linking to https://github.com/googleapis/librarian/issues/564